mhs/mhs/common/messages/ebxml_envelope.py

```diff
@@ -68,67 +76,49 @@ def serialize(self) -> Tuple[str, str]:
-        EbxmlEnvelope._add_flag_if_present(extracted_values, SYNC_REPLY,
-                                           EbxmlEnvelope._extract_ebxml_value(xml_tree, "SyncReply"))
-        EbxmlEnvelope._add_flag_if_present(extracted_values, ACK_REQUESTED,
-                                           EbxmlEnvelope._extract_ebxml_value(xml_tree, "AckRequested"))
-        EbxmlEnvelope._extract_attribute(xml_tree, "AckRequested", SOAP_NAMESPACE, "actor", extracted_values,
-                                         ACK_SOAP_ACTOR)
+        for element_to_extract in EbxmlEnvelope._elements_to_extract_when_parsing:
+            EbxmlEnvelope._add_if_present(extracted_values, element_to_extract['name'],
+                                          EbxmlEnvelope._extract_ebxml_text_value(xml_tree,
+                                                                                  element_to_extract['element_name'],
+                                                                                  parent=element_to_extract['parent'],
+                                                                                  required=True))
 
         return extracted_values
 
     @staticmethod
-    def _path_to_ebxml_element(name, parent=None):
+    def _path_to_ebxml_element(name: str, parent: str = None) -> str:
         path = ".//"
 
         if parent is not None:
-            path += EBXML_NAMESPACE + ":" + parent + "/"
+            path += f"{EBXML_NAMESPACE}:{parent}/"
 
-        path += EBXML_NAMESPACE + ":" + name
+        path += f"{EBXML_NAMESPACE}:{name}"
 
         return path
 
     @staticmethod
-    def _extract_ebxml_value(xml_tree, element_name, parent=None):
+    def _extract_ebxml_value(xml_tree: ElementTree.Element, element_name: str, parent: str = None,
+                             required: bool = False) -> Optional[ElementTree.Element]:
         xpath = EbxmlEnvelope._path_to_ebxml_element(element_name, parent=parent)
-        return xml_tree.find(xpath, namespaces=NAMESPACES)
+        value = xml_tree.find(xpath, namespaces=NAMESPACES)
+        if value is None and required:
+            logger.error('0002', "Weren't able to find required element {xpath} during parsing of EbXML message.",
+                         {'xpath': xpath})
+            raise EbXmlParsingError(f"Weren't able to find required element {xpath} during parsing of EbXML message")
+        return value
 
     @staticmethod
-    def _extract_ebxml_text_value(xml_tree, element_name, parent=None):
-        value = EbxmlEnvelope._extract_ebxml_value(xml_tree, element_name, parent)
+    def _extract_ebxml_text_value(xml_tree: ElementTree.Element, element_name: str, parent: str = None,
+                                  required: bool = False) -> Optional[str]:
+        value = EbxmlEnvelope._extract_ebxml_value(xml_tree, element_name, parent, required)
         text = None
 
         if value is not None:
```

Comment:
Here and [here](https://github.com/nhsconnect/integration-adaptors/pull/46/files#diff-f28d44ca274c270cb3ef934a33e38dd2R150), the unittests I've got don't test when the if statement is False. The reason why is that atm we don't parse any templates with optional text values. I considered creating some sort of contrived test for this scenario, but `TestEbxmlEnvelope` is used as a base class, so I'd have to somehow fiddle with that. I could maybe change the interface, but that feels a bit odd as well as sometime in the future we may want to optionally extract a text value.

Anyone got any thoughts as to what they'd prefer I do?

Reply:
Maybe I don't understand fully what's going on here but would it suffice just to test this method and `parse_message` in isolation to ensure it can handle optional values? This should give us _some_ confidence it handles optionals without having to build a contrived test? Then maybe at a later date when we have optional text values we can build more suitable tests if needed